jobs/joint_genotyping.py

```diff
@@ -0,0 +1,735 @@
+        j.depends_on(*depends_on)
+
+    # The Broad: testing has shown that the multithreaded reader initialization
+    # does not scale well beyond 5 threads, so don't increase beyond that.
+    nthreads = 5
+
+    # The Broad: The memory setting here is very important and must be several
+    # GiB lower than the total memory allocated to the VM because this tool uses
+    # a significant amount of non-heap memory for native libraries.
+    xms_gb = 8
+    xmx_gb = 25
+
+    STANDARD.set_resources(j, nthreads=nthreads, mem_gb=xmx_gb + 1)
+
+    params = [
+        # The Broad: We've seen some GenomicsDB performance regressions related
+        # to intervals, so we're going to pretend we only have a single interval
+        # using the --merge-input-intervals arg. There's no data in between since we
+        # didn't run HaplotypeCaller over those loci, so we're not wasting any compute
+        '--merge-input-intervals',
+        '--consolidate',
+        # The batch_size value was carefully chosen here as it is the optimal value for
+        # the amount of memory allocated within the task; please do not change it
+        # without consulting the Hellbender (GATK engine) team!
+        '--batch-size',
+        '50',
+        # Using cloud + --consolidate causes a TileDB error:
+        # https://github.com/broadinstitute/gatk/issues/7653
+        # However disabling --consolidate decreases performance of reading, cause
+        # GenotypeGVCFs to run forever. So instead, using non-cloud GenomicsDB.
+    ]
+
+    j.command(
+        command(
+            f"""\
+
+    echo "{msg}"
+
+    {rm_cmd}
+
+    gatk --java-options "-Xms{xms_gb}g -Xmx{xmx_gb}g" \
+    GenomicsDBImport \\
+    {genomicsdb_param} \\
+    -L {intervals} \\
+    --sample-name-map {sample_map} \\
+    --reader-threads {nthreads} \\
+    {" ".join(params)}
+    """,
+            monitor_space=True,
+            setup_gcp=True,
+        )
+    )
+    return j, sample_names_will_be_in_db
+
+
+def _add_joint_genotyper_job(
+    b: hb.Batch,
+    genomicsdb_path: Path,
+    overwrite: bool,
+    number_of_samples: int,
+    scatter_count: int,
+    interval: hb.Resource | None = None,
+    output_vcf_path: Path | None = None,
+    tool: JointGenotyperTool = JointGenotyperTool.GnarlyGenotyper,
+    job_attrs: dict | None = None,
+) -> tuple[Job | None, hb.ResourceGroup]:
+    """
+    Runs GATK GnarlyGenotyper or GenotypeGVCFs on a combined_gvcf VCF bgzipped file,
+    pre-called with HaplotypeCaller.
+
+    GenotypeGVCFs is a standard GATK joint-genotyping tool.
+
+    GnarlyGenotyper is an experimental GATK joint-genotyping tool that performs
+    "quick and dirty" joint genotyping on large cohorts, of GVCFs post-processed
+    with ReblockGVCF.
+
+    HaplotypeCaller must be used with `-ERC GVCF` or `-ERC BP_RESOLUTION` to add
+    genotype likelihoods.
+
+    ReblockGVCF must be run to add all the annotations necessary for VQSR:
+    QUALapprox, VarDP, RAW_MQandDP.
```

Comment:
I'm still puzzled why that's part of `ReblockGVCF`... ;)

Reply:
GATK acts in mysterious ways.

Reply:
It would be amazing though in some future to just switch to Hail Combiner for this